tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraContactAutocompleteMethodContract.scala

```diff
@@ -0,0 +1,692 @@
+      .body(request)
+    .when
+      .post
+    .`then`
+      .log().ifValidationFails()
+      .statusCode(SC_OK)
+      .contentType(JSON)
+      .extract
+      .body
+      .asString
+
+    assertThatJson(response)
+      .isEqualTo(
+        s"""{
+           |    "sessionState": "${SESSION_STATE.value}",
+           |    "methodResponses": [
+           |        [
+           |            "error",
+           |            {
+           |                "type": "invalidArguments",
+           |                "description": "{\\"errors\\":[{\\"path\\":\\"obj.filter\\",\\"messages\\":[\\"error.path.missing\\"]}]}"
+           |            },
+           |            "c1"
+           |        ]
+           |    ]
+           |}
+           |""".stripMargin)
+  }
+
+  @Test
+  def shouldReturnInvalidArgumentsWhenInvalidFilterCondition(): Unit = {
+    val request =
+      s"""{
+         |  "using": ["urn:ietf:params:jmap:core", "com:linagora:params:jmap:autocomplete"],
+         |  "methodCalls": [[
+         |    "TMailContact/autocomplete",
+         |    {
+         |      "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |      "filter": {
+         |        "unsupported_option": "blabla",
+         |        "text":"obi"
+         |      }
+         |    },
+         |    "c1"]]
+         |}""".stripMargin
+
+    val response =  `given`
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .body(request)
+    .when
+      .post
+    .`then`
+      .log().ifValidationFails()
+      .statusCode(SC_OK)
+      .contentType(JSON)
+      .extract
+      .body
+      .asString
+
+    assertThatJson(response)
+      .isEqualTo(
+        s"""{
+           |    "sessionState": "${SESSION_STATE.value}",
+           |    "methodResponses": [
+           |        [
+           |            "error",
+           |            {
+           |                "type": "invalidArguments",
+           |                "description": "{\\"errors\\":[{\\"path\\":\\"obj.filter\\",\\"messages\\":[\\"These '[unsupported_option]' was unsupported filter options\\"]}]}"
+           |            },
+           |            "c1"
+           |        ]
+           |    ]
+           |}
+           |""".stripMargin)
+  }
+
+  @Test
+  def limitShouldBeTakenIntoAccount(server: GuiceJamesServer): Unit = {
+    val contactCreatedA = server.getProbe(classOf[JmapGuiceContactAutocompleteProbe])
+      .index(bobAccountId, contactA)
+    server.getProbe(classOf[JmapGuiceContactAutocompleteProbe])
+      .index(DOMAIN, contactB)
+
+    val request =
+      s"""{
+         |  "using": ["urn:ietf:params:jmap:core", "com:linagora:params:jmap:autocomplete"],
+         |  "methodCalls": [[
+         |    "TMailContact/autocomplete",
+         |    {
+         |      "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |      "filter": {"text":"obi"},
+         |      "limit": 1
```

Comment:
Q: 
When the client set `limit` parameter, the response will NOT contain `limit`?
Is it correct?

Reply:
Almost correct. The response will not contain the limit if it is lower than the default set by the server (the one applied then).

If the limit set by client is too high (like 300 while server limit is 256), the used limit has to be specified in the response

Reply:
Yes as the limit is the one set by the client.

The limit is only answered when the value is different from the one supplied by the client.